src/main/java/io/cdap/plugin/zendesk/connector/ZendeskConnector.java

```diff
@@ -0,0 +1,210 @@
+    this.config = config;
+  }
+
+  @Override
+  public void test(ConnectorContext connectorContext) throws ValidationException {
+    FailureCollector collector = connectorContext.getFailureCollector();
+    config.validateConnectionParameters(collector);
+  }
+
+  @Override
+  public BrowseDetail browse(ConnectorContext connectorContext, BrowseRequest browseRequest) {
+    BrowseDetail.Builder browseDetailBuilder = BrowseDetail.builder();
+    List<BrowseEntity> entityList = Arrays.stream(ObjectType.values()).filter(t -> t.isSamplingAllowed())
+      .map(object -> {
+        BrowseEntity.Builder entity = (BrowseEntity.builder(object.getObjectName(),
+          object.getObjectName(),
+          ENTITY_TYPE_OBJECT).canBrowse(false).canSample(true));
+        return entity.build();
+      }).collect(Collectors.toList());
+    return browseDetailBuilder.addEntities(entityList).setTotalCount(entityList.size()).build();
+  }
+
+  @Override
+  public ConnectorSpec generateSpec(ConnectorContext connectorContext, ConnectorSpecRequest connectorSpecRequest) {
+    FailureCollector collector = connectorContext.getFailureCollector();
+    ConnectorSpec.Builder specBuilder = ConnectorSpec.builder();
+    Map<String, String> properties = new HashMap<>();
+    properties.put(ConfigUtil.NAME_USE_CONNECTION, "true");
+    properties.put(ConfigUtil.NAME_CONNECTION, connectorSpecRequest.getConnectionWithMacro());
+    String objectName = connectorSpecRequest.getPath();
+    if (objectName != null) {
+      properties.put(ZendeskBatchSourceConfig.PROPERTY_OBJECTS_TO_PULL, objectName);
+      properties.put(Constants.Reference.REFERENCE_NAME, ReferenceNames.cleanseReferenceName(objectName));
+    }
+    Schema schema = ObjectType.fromString(objectName, collector).getObjectSchema();
+    specBuilder.setSchema(schema);
+    return specBuilder.addRelatedPlugin(
+      new PluginSpec(ZendeskBatchSource.NAME, BatchSource.PLUGIN_TYPE, properties)).build();
+  }
+
+  @Override
+  public List<StructuredRecord> sample(ConnectorContext connectorContext, SampleRequest sampleRequest)
+    throws IOException {
+    FailureCollector collector = connectorContext.getFailureCollector();
+    List<StructuredRecord> recordList = new ArrayList<>();
+    objectType = ObjectType.fromString(sampleRequest.getPath(), collector);
+    Schema schema = objectType.getObjectSchema();
+    for (String record : getResponseAsMap()) {
+      recordList.add(StructuredRecordStringConverter.fromJsonString(record, schema));
+    }
+    return recordList;
+  }
+
+  private List<String> getResponseAsMap() throws IOException {
+    subdomain = config.getSubdomains().stream().findFirst().get();
+    String baseUrl = String.format(ZendeskConstants.BASE_URL, subdomain, objectType.getApiEndpoint());
+    if (objectType.isBatch()) {
+      // start time = current date - 30 days
+      // start time is set like this because for some objects start time is mandatory to fetch the data according to
```

Comment:
can we set a limit (= sampleRequest.getLimit()) on the number of records fetched in addition to the start time?

Reply:
Here single request is made to fetch the data which fetches 1000 records only at a time. So addition of sampleRequest.getLimit() is not needed here.

Reply:
Can you fetch number of records =  sampleRequest.getLimit() instead of 1000?

Reply:
setting a limit to fetch certain number of records is not available with time based pagination APIs and not working properly even after they it is mentioned in Zendesk APIs.
https://developer.zendesk.com/api-reference/introduction/pagination/ 
So it is advisable not to use it.

Reply:
Can we use [cursor pagination](https://developer.zendesk.com/api-reference/introduction/pagination/#using-cursor-pagination) and specify `page[size]` parameter to specify the number of items to return?
Alternatively, looks like [offset pagination](https://developer.zendesk.com/api-reference/introduction/pagination/#using-offset-pagination) supports a `per_page` that specifies the max number of records per page.
Can either of these options be used instead of fetching 30 days worth of data? I'm mostly worried about OOM issues if we fetch too much data.
Also @Shubhangi-cs mentioned that only 1000 records are fetched at a time with time based pagination. Can you point me to where that's documented?

Reply:
Cursor pagination APIs has data loss issues because of which we have removed it from the code for the objects which support cursor pagination(only 2). We will not face OOM issues because we are not fetching all the data for 30 days instead it will make a single API call which can fetch maximum 1000 records. Zendesk documentation is bit flawed as of now, they are lot of inconsistencies over there and we have raised tickets for that. We have tried both page[size]=1000 and per_page=1000 in URL as query params and they are not working as expected.

You can refer below link where they have mentioned maximum records will be 1000.
https://developer.zendesk.com/documentation/ticketing/managing-tickets/using-the-incremental-export-api/#time-based-incremental-exports

The only reason we are giving 30 days is that it is mandatory field but we are not extracting the whole data as mentioned above.

Reply:
Thanks for the explanation! Please add this as a comment, make sure to include the following details:
- We can't specify the number of records to read due to bugs in Zendesk API
- Workaround is time based pagination where we will read a max of 1000 records (add link to doc).